### ISSUE TITLE
Erroneous error!  IDE1007 The name 'var' does not exist in the current context

### ISSUE THREAD

**Issue by joshlang** [dotnet/roslyn#38947]
Using VS 2019:  16.4.0 Preview 1.0

Consider this code:
```
using System;

namespace ConsoleApp3
{
    class Program
    {
        static void Main(string[] args)
        {
            Action a = () =>
            {
                var (x, y) = (4, 5);
#warning Hi
            };
        }
    }
}
```

It's fine.  But the error window shows: `IDE1007 The name 'var' does not exist in the current context` and a red squiggly appears under the word 'var'.

It compiles without any errors, though.

Any of the following will remove the IDE1007 error:
- Don't deconstruct the tuple.  `var x = (4, 5)`
- Move the offending code into void Main() instead of in a lambda.
- Remove the #warning (The problem only exists if there are any warnings in the lambda's scope, whether naturally or by creating one with #warning)

**Comment by joshlang** [comment]
Erroneous spelling of Erroneous, too.

**Comment by snhardin** [comment]
Hello! I'd like to take a stab at this if possible. If anybody has any pointers as to where I should look first, I'd very much appreciate it.

**Comment by sharwell** [comment]
@snhardin This diagnostic is reported by [`CSharpUnboundIdentifiersDiagnosticAnalyzer`](https://github.com/dotnet/roslyn/blob/master/src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUnboundIdentifiersDiagnosticAnalyzer.cs). If I was starting on this, I would create a new unit test in [`GenerateTypeTests`](https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs) and set a breakpoint on the line where the diagnostic is reported:

https://github.com/dotnet/roslyn/blob/cf104c5204f51d75e79c93935b8bb96b9c8d3796/src/Features/Core/Portable/Diagnostics/Analyzers/UnboundIdentifiersDiagnosticAnalyzerBase.cs#L83

**Comment by snhardin** [comment]
Thanks for the pointer @sharwell. Indeed, the block you pointed out seems to be the important one.
https://github.com/dotnet/roslyn/blob/cf104c5204f51d75e79c93935b8bb96b9c8d3796/src/Features/Core/Portable/Diagnostics/Analyzers/UnboundIdentifiersDiagnosticAnalyzerBase.cs#L74-L84

I did some investigation, and it seems that `GetSymbolInfo()` returns `SymbolInfo.None` for the `var` tuple declaration. This seems to me to be intended functionality, since hovering over a `var` for a tuple declaration like this provides no hints on what types are being declared. I'm thinking that an additional check needs to happen on the same line as `IsNameOf()`.

I have a proof of concept that I think is working, although I'd like some advice on how you guys would want it actually implemented. I'll submit a WIP PR soon and then you guys will be free to tell me that I'm doing it all wrong in the code review. 🙂

**Comment by Logopher** [comment]
> snhardin linked a pull request that will close this issue on Oct 16, 2019

???

I'm seeing this problem currently in VS 2019 16.7.1. The test case fails and is fixed by removing the warning.

**Comment by CyrusNajmabadi** [comment]
> I'm seeing this problem currently in VS 2019 16.7.1.

The PR fixing this has never been merged in.

**Comment by snhardin** [comment]
Yes this was never merged in or reviewed. I've long since lost interest in getting this in. Someone else is welcome to use my code or develop their own fix.
